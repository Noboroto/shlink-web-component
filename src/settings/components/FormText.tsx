import type { FC, PropsWithChildren } from 'react';

export const FormText: FC<PropsWithChildren> = ({ children }) => (
  <small className="form-text text-muted d-block">{children}</small>
);
