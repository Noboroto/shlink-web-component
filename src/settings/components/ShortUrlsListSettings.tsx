import { LabeledFormGroup, OrderingDropdown, SimpleCard } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import type { ShortUrlsListSettings as ShortUrlsSettings } from '..';
import { useSetting } from '..';

export type ShortUrlsListSettingsProps = {
  updateShortUrlsListSettings: (settings: ShortUrlsSettings) => void;
  defaultOrdering: NonNullable<ShortUrlsSettings['defaultOrdering']>;
};

const SHORT_URLS_ORDERABLE_FIELDS = {
  dateCreated: 'Created at',
  shortCode: 'Short URL',
  longUrl: 'Long URL',
  title: 'Title',
  visits: 'Visits',
};

export const ShortUrlsListSettings: FC<ShortUrlsListSettingsProps> = (
  { updateShortUrlsListSettings, defaultOrdering },
) => {
  const shortUrlsList = useSetting('shortUrlsList');

  return (
    <SimpleCard title="Short URLs list" className="h-100">
      <LabeledFormGroup noMargin label="Default ordering for short URLs list:">
        <OrderingDropdown
          items={SHORT_URLS_ORDERABLE_FIELDS}
          order={shortUrlsList?.defaultOrdering ?? defaultOrdering}
          onChange={(field, dir) => updateShortUrlsListSettings({ defaultOrdering: { field, dir } })}
        />
      </LabeledFormGroup>
    </SimpleCard>
  );
};
