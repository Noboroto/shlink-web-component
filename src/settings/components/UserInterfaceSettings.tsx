import { faMoon, faSun } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import type { Theme } from '@shlinkio/shlink-frontend-kit';
import { getSystemPreferredTheme, SimpleCard, ToggleSwitch } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { useMemo } from 'react';
import { useSetting } from '..';
import type { UiSettings } from '../types';

interface UserInterfaceProps {
  updateUiSettings: (settings: UiSettings) => void;

  /* Test seam */
  _matchMedia?: typeof window.matchMedia;
}

export const UserInterfaceSettings: FC<UserInterfaceProps> = ({ updateUiSettings, _matchMedia }) => {
  const ui = useSetting('ui');
  const currentTheme = useMemo(() => ui?.theme ?? getSystemPreferredTheme(_matchMedia), [ui?.theme, _matchMedia]);

  return (
    <SimpleCard title="User interface" className="h-100">
      <FontAwesomeIcon icon={currentTheme === 'dark' ? faMoon : faSun} className="float-end mt-1" />
      <ToggleSwitch
        checked={currentTheme === 'dark'}
        onChange={(useDarkTheme) => {
          const theme: Theme = useDarkTheme ? 'dark' : 'light';
          updateUiSettings({ ...ui, theme });
        }}
      >
        Use dark theme.
      </ToggleSwitch>
    </SimpleCard>
  );
};
