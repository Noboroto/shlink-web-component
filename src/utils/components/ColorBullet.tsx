import './ColorBullet.scss';

interface ColorBulletProps {
  color: string;
}

export const ColorBullet = ({ color }: ColorBulletProps) => (
  <div style={{ backgroundColor: color }} className="color-bullet" />
);
