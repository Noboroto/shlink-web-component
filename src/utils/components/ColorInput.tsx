import { useElementRef } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { Input, InputGroup } from 'reactstrap';
import type { ColorPickerProps } from './ColorPicker';
import { ColorPicker } from './ColorPicker';

export const ColorInput: FC<Omit<ColorPickerProps, 'className'>> = ({ color, onChange }) => {
  const colorPickerRef = useElementRef<HTMLInputElement>();

  return (
    <InputGroup>
      <ColorPicker color={color} onChange={onChange} className="input-group-text" ref={colorPickerRef} />
      <Input value={color} readOnly onClick={() => colorPickerRef.current?.click()} />
    </InputGroup>
  );
};
