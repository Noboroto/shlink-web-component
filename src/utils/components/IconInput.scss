@import '../mixins/vertical-align';
@import '../../../node_modules/@shlinkio/shlink-frontend-kit/dist/base';

.icon-input-container {
  position: relative;
}

.icon-input-container__input {
  padding-right: 35px !important;
}

.icon-input-container__input:not(:disabled) {
  background-color: var(--primary-color) !important;
}

.card .icon-input-container__input:not(:disabled),
.dropdown .icon-input-container__input:not(:disabled) {
  background-color: var(--input-color) !important;
}

.icon-input-container__icon {
  @include vertical-align();

  right: .75rem;
  cursor: pointer;
}
