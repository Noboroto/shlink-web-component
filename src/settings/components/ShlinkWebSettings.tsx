import { mergeDeepRight } from '@shlinkio/data-manipulation';
import { NavPillItem, NavPills } from '@shlinkio/shlink-frontend-kit';
import type { FC, ReactNode } from 'react';
import { useCallback } from 'react';
import { Navigate, Route, Routes } from 'react-router-dom';
import type { DeepPartial } from '../../utils/types';
import { SettingsProvider } from '..';
import type { RealTimeUpdatesSettings, Settings, ShortUrlsListSettings } from '../types';
import { RealTimeUpdatesSettings as RealTimeUpdates } from './RealTimeUpdatesSettings';
import { ShortUrlCreationSettings as ShortUrlCreation } from './ShortUrlCreationSettings';
import { ShortUrlsListSettings as ShortUrlsList } from './ShortUrlsListSettings';
import { TagsSettings as Tags } from './TagsSettings';
import { UserInterfaceSettings } from './UserInterfaceSettings';
import { VisitsSettings as Visits } from './VisitsSettings';

export type ShlinkWebSettingsProps = {
  settings: Settings;
  defaultShortUrlsListOrdering: NonNullable<ShortUrlsListSettings['defaultOrdering']>;
  updateSettings: (settings: Settings) => void;
};

const SettingsSections: FC<{ items: ReactNode[] }> = ({ items }) => (
  <>
    {items.map((child, index) => <div key={index} className="mb-3">{child}</div>)}
  </>
);

export const ShlinkWebSettings: FC<ShlinkWebSettingsProps> = (
  { settings, updateSettings, defaultShortUrlsListOrdering },
) => {
  const updatePartialSettings = useCallback(
    (partialSettings: DeepPartial<Settings>) => updateSettings(mergeDeepRight(settings, partialSettings)),
    [settings, updateSettings],
  );
  const toggleRealTimeUpdates = useCallback(
    (enabled: boolean) => updatePartialSettings({ realTimeUpdates: { enabled } }),
    [updatePartialSettings],
  );
  const setRealTimeUpdatesInterval = useCallback(
    (interval: number) => updatePartialSettings({ realTimeUpdates: { interval } as RealTimeUpdatesSettings }),
    [updatePartialSettings],
  );
  const updateSettingsProp = useCallback(
    <Prop extends keyof Settings>(prop: Prop, value: Settings[Prop]) => updatePartialSettings({ [prop]: value }),
    [updatePartialSettings],
  );

  return (
    <SettingsProvider value={settings}>
      <NavPills className="mb-3">
        <NavPillItem to="general">General</NavPillItem>
        <NavPillItem to="short-urls">Short URLs</NavPillItem>
        <NavPillItem to="other-items">Other items</NavPillItem>
      </NavPills>

      <Routes>
        <Route
          path="general"
          element={(
            <SettingsSections
              items={[
                <UserInterfaceSettings updateUiSettings={(v) => updateSettingsProp('ui', v)} />,
                <RealTimeUpdates
                  toggleRealTimeUpdates={toggleRealTimeUpdates}
                  setRealTimeUpdatesInterval={setRealTimeUpdatesInterval}
                />,
              ]}
            />
          )}
        />
        <Route
          path="short-urls"
          element={(
            <SettingsSections
              items={[
                <ShortUrlCreation updateShortUrlCreationSettings={(v) => updateSettingsProp('shortUrlCreation', v)} />,
                <ShortUrlsList
                  defaultOrdering={defaultShortUrlsListOrdering}
                  updateShortUrlsListSettings={(v) => updateSettingsProp('shortUrlsList', v)}
                />,
              ]}
            />
          )}
        />
        <Route
          path="other-items"
          element={(
            <SettingsSections
              items={[
                <Tags updateTagsSettings={(v) => updateSettingsProp('tags', v)} />,
                <Visits updateVisitsSettings={(v) => updateSettingsProp('visits', v)} />,
              ]}
            />
          )}
        />
        <Route path="*" element={<Navigate replace to="general" />} />
      </Routes>
    </SettingsProvider>
  );
};
