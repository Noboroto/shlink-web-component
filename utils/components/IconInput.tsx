import type { IconProp } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classNames from 'classnames';
import type { FC } from 'react';
import type { InputProps } from 'reactstrap';
import { Input } from 'reactstrap';
import { useElementRef } from '../../../shlink-frontend-kit/src';
import './IconInput.scss';

type IconInputProps = InputProps & {
  icon: IconProp;
};

export const IconInput: FC<IconInputProps> = ({ icon, className, ...rest }) => {
  const ref = useElementRef<HTMLInputElement>();
  const classes = classNames('icon-input-container__input', className);

  return (
    <div className="icon-input-container">
      <Input className={classes} innerRef={ref} {...rest} />
      <FontAwesomeIcon
        icon={icon}
        fixedWidth
        className="icon-input-container__icon"
        onClick={() => ref.current?.focus()}
      />
    </div>
  );
};
