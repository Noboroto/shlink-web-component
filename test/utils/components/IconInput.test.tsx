import type { IconProp } from '@fortawesome/fontawesome-svg-core';
import { faAppleAlt, faCalendar, faTable } from '@fortawesome/free-solid-svg-icons';
import { screen } from '@testing-library/react';
import { IconInput } from '../../../src/utils/components/IconInput';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { renderWithEvents } from '../../__helpers__/setUpTest';

describe('<IconInput />', () => {
  const setUp = (icon: IconProp, placeholder?: string) => renderWithEvents(
    <IconInput icon={icon} placeholder={placeholder} />,
  );

  it('passes a11y checks', () => checkAccessibility(setUp(faCalendar, 'some placeholder')));

  it.each([faCalendar, faAppleAlt, faTable])('displays provided icon', (icon) => {
    const { container } = setUp(icon);
    expect(container).toMatchSnapshot();
  });

  it('focuses input on icon click', async () => {
    const { user } = setUp(faCalendar, 'foo');

    expect(screen.getByPlaceholderText('foo')).not.toHaveFocus();
    await user.click(screen.getByRole('img', { hidden: true }));
    expect(screen.getByPlaceholderText('foo')).toHaveFocus();
  });
});
