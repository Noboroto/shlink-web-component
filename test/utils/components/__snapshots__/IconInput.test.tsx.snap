// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<IconInput /> > displays provided icon 1`] = `
<div>
  <div
    class="icon-input-container"
  >
    <input
      class="icon-input-container__input form-control"
      type="text"
    />
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-calendar fa-fw icon-input-container__icon"
      data-icon="calendar"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`<IconInput /> > displays provided icon 2`] = `
<div>
  <div
    class="icon-input-container"
  >
    <input
      class="icon-input-container__input form-control"
      type="text"
    />
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-apple-whole fa-fw icon-input-container__icon"
      data-icon="apple-whole"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M224 112c-8.8 0-16-7.2-16-16V80c0-44.2 35.8-80 80-80h16c8.8 0 16 7.2 16 16V32c0 44.2-35.8 80-80 80H224zM0 288c0-76.3 35.7-160 112-160c27.3 0 59.7 10.3 82.7 19.3c18.8 7.3 39.9 7.3 58.7 0c22.9-8.9 55.4-19.3 82.7-19.3c76.3 0 112 83.7 112 160c0 128-80 224-160 224c-16.5 0-38.1-6.6-51.5-11.3c-8.1-2.8-16.9-2.8-25 0c-13.4 4.7-35 11.3-51.5 11.3C80 512 0 416 0 288z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;

exports[`<IconInput /> > displays provided icon 3`] = `
<div>
  <div
    class="icon-input-container"
  >
    <input
      class="icon-input-container__input form-control"
      type="text"
    />
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-table fa-fw icon-input-container__icon"
      data-icon="table"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M64 256V160H224v96H64zm0 64H224v96H64V320zm224 96V320H448v96H288zM448 256H288V160H448v96zM64 32C28.7 32 0 60.7 0 96V416c0 35.3 28.7 64 64 64H448c35.3 0 64-28.7 64-64V96c0-35.3-28.7-64-64-64H64z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
